server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailConverter.java

```diff
@@ -0,0 +1,176 @@
+            .sender(deletedMessage.getSender())
+            .recipients(deletedMessage.getRecipients())
+            .mimeMessage(new MimeMessage(Session.getDefaultInstance(new Properties()), deletedMessage.getContent().get()))
+            .attribute(OWNER_ATTRIBUTE_NAME.withValue(SerializableUser.toAttributeValue(deletedMessage.getOwner())))
+            .attribute(HAS_ATTACHMENT_ATTRIBUTE_NAME.withValue(AttributeValue.of(deletedMessage.hasAttachment())))
+            .attribute(DELIVERY_DATE_ATTRIBUTE_NAME.withValue(SerializableDate.toAttributeValue(deletedMessage.getDeliveryDate())))
+            .attribute(DELETION_DATE_ATTRIBUTE_NAME.withValue(SerializableDate.toAttributeValue(deletedMessage.getDeletionDate())))
+            .attribute(ORIGIN_MAILBOXES_ATTRIBUTE_NAME.withValue(AttributeValue.of(serializedMailboxIds(deletedMessage))))
+            .attribute(SUBJECT_ATTRIBUTE_VALUE.withValue(subjectAttributeValue(deletedMessage)))
+            .build();
+    }
+
+    DeletedMessage fromMail(Mail mail) {
+        MemoizedSupplier<byte[]> byteContent = new MemoizedSupplier<>(Throwing.supplier(
+            () -> MimeMessageUtil.asBytes(mail.getMessage())));
+
+        return DeletedMessage.builder()
+            .messageId(messageIdFactory.fromString(mail.getName()))
+            .originMailboxes(retrieveMailboxIds(mail))
+            .user(retrieveOwner(mail))
+            .deliveryDate(retrieveDate(mail, DELIVERY_DATE_ATTRIBUTE_NAME))
+            .deletionDate(retrieveDate(mail, DELETION_DATE_ATTRIBUTE_NAME))
+            .sender(mail.getMaybeSender())
+            .recipients(mail.getRecipients())
+            .content(() -> new ByteArrayInputStream(byteContent.get()))
+            .hasAttachment(retrieveHasAttachment(mail))
+            .subject(retrieveSubject(mail))
+            .build();
+    }
+
+    private ImmutableList<AttributeValue<?>> serializedMailboxIds(DeletedMessage deletedMessage) {
+        return deletedMessage.getOriginMailboxes().stream()
+            .map(MailboxId::serialize)
+            .map(AttributeValue::of)
+            .collect(Guavate.toImmutableList());
+    }
+
+    private AttributeValue<Optional<AttributeValue<String>>> subjectAttributeValue(DeletedMessage deletedMessage) {
+        return AttributeValue.of(deletedMessage.getSubject().map(AttributeValue::of));
+    }
+
+    private Optional<String> retrieveSubject(Mail mail) {
+        return mail.getAttribute(SUBJECT_ATTRIBUTE_VALUE)
+                .map(Attribute::getValue)
+                .flatMap(attributeValue -> attributeValue.valueAs(Optional.class))
+                .map(this::retrieveSubject)
+                .orElseThrow(() -> new IllegalArgumentException("mail should have a 'subject' attribute being of type 'Optional<String>"));
+    }
+
+    private Optional<String> retrieveSubject(Optional<?> optional) {
+        return optional.map(this::castSubjectToString);
+    }
+
+    private String castSubjectToString(Object x) {
```

Comment:
can you rename `x` ?

Reply:
Well at this stage, this "untyped object" clearly needs additional checks to enforce clearly what it is hence I do believe a cryptic name `x` is quite good for varialbes having cryptic type.

Suggestions?

Reply:
object? unknownObject? untypedObject?